### PR TITLE
Profiling based refactoring in heat balance functions

### PR DESCRIPTION
**Pull request by xuanluo113**
Pull request overview
---------------------
 - Refactoring of `SetWindSpeedAt`, `SetSurfacetempAt` at surface level.
 - Refactoring of `ReportSurfaceHeatBalance` and `CalcThermalResilience`.
 - Refactoring of the CTF term loop in`InitSurfaceHeatBalance`
 - Refactoring of updating some meter values updates in Outputprocessor.
 - Diffs expected for SurfaceZonePropTest_LocalEnv (see https://github.com/NREL/EnergyPlus/pull/8949#issuecomment-909484150).
 - Performance improvement 1.24% as of e546b374fb81b1dc5c84e3b3977e56e68d399530

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

**Comment by xuanluo113** [comment]
@nealkruis @mjwitte @JasonGlazer I'll stop making more changes to this branch and mark it ready for review, in case it continues to stepping on other branches. Still waiting for the CI, but previous (washed out) CI results were normal. Wonder if you can help to review this. The main changes are explained in the PR overview. Thanks!

File: src/EnergyPlus/DataSurfaces.cc [hunk 1]
```diff
@@ -661,8 +588,28 @@ void SurfaceData::set_representative_surface(EnergyPlusData &state, const int Su
 
 void SetSurfaceOutBulbTempAt(EnergyPlusData &state)
 {
-    for (int SurfNum = 1; SurfNum <= state.dataSurface->TotSurfaces; SurfNum++) {
-        SetOutBulbTempAt(state, SurfNum);
+    if (state.dataEnvrn->SiteTempGradient == 0.0) {
```
Comment (amirroth) [inline src/EnergyPlus/DataSurfaces.cc:581]: Are these needed for all surfaces or just exterior surfaces? Does it make sense to create a vector of all exterior surfaces and just iterate over that?

File: src/EnergyPlus/HeatBalanceSurfaceManager.cc [hunk 2]
```diff
@@ -5251,33 +5263,31 @@ void CalcThermalResilience(EnergyPlusData &state)
...
-    Real64 c7 = .00122874;
-    Real64 c8 = .00085282;
-    Real64 c9 = -.00000199;
-
     // Calculate Heat Index and Humidex.
     // The heat index equation set is fit to Fahrenheit units, so the zone air temperature values are first convert to F,
     // then heat index is calculated and converted back to C.
     if (state.dataHeatBalSurfMgr->reportVarHeatIndex || state.dataOutRptTab->displayThermalResilienceSummary) {
+        // Constance for heat index regression equation of Rothfusz.
+        Real64 const c1 = -42.379;
```
Comment (amirroth) [inline src/EnergyPlus/HeatBalanceSurfaceManager.cc]: Not sure if it actually makes a difference given the scope, but can probably make these `constexpr`.

File: src/EnergyPlus/OutputProcessor.cc [hunk 3]
```diff
@@ -2587,106 +2461,83 @@ namespace OutputProcessor {
...
+            if (op->isFinalYear) {
+                if (op->EnergyMeters(Meter).TSValue < op->EnergyMeters(Meter).FinYrSMMinVal) {
+                    op->EnergyMeters(Meter).FinYrSMMinVal = op->EnergyMeters(Meter).TSValue;
+                    op->EnergyMeters(Meter).FinYrSMMinValDate = TimeStamp;
+                }
+            }
+        }
+        for (int Meter = 1; Meter <= op->NumEnergyMeters; ++Meter) {
+            op->MeterValue(Meter) = 0.0; // Ready for next update
+        }
```
Comment (amirroth) [inline src/EnergyPlus/OutputProcessor.cc:2542]: Nice!!

**Comment by mjwitte** [comment]
@xuanluo113 There are big diffs in SurfaceZonePropTest_LocalEnv that need fixing or explanation.

**Review by JasonGlazer — COMMENTED** [review]
For the updated functions, have you looked at how often consecutive calls use the same argument values? If this happens fairly often, it might make sense to store the previous call argument values and result and check if the new call uses the same argument values.

**Comment by xuanluo113** [comment]
@mjwitte The big diffs in `SurfaceZonePropTest_LocalEnv` are expected. The diff only exists in reporting surface outdoor air wind speed for four **Exterior** surfaces with **NoWind** as a user input for the field of "Wind Exposure".

[decent_ci:test_result:message] ESO big diffs:
Screen Shot 2021-08-31 at 11 12 39 AM

Screen Shot 2021-08-31 at 11 08 13 AM

In the refactored code, we don't calculate the surface outdoor wind speed with **NoWind** indicated because it's not used, while previously we do.

**Comment by mjwitte** [comment]
@xuanluo113 Thanks for explaining the diffs. This branch has conflicts now.

**Comment by amirroth** [comment]
It's very windy in Maine.

File: src/EnergyPlus/HeatBalanceSurfaceManager.cc (hunk 2 above)
Comment (xuanluo113) [reply src/EnergyPlus/HeatBalanceSurfaceManager.cc]: Okay. Done.

File: src/EnergyPlus/DataSurfaces.cc (hunk 1 above)
Comment (xuanluo113) [reply src/EnergyPlus/DataSurfaces.cc:581]: I tried applying the calculation to exterior surfaces only and it creates diffs for some test files with natural ventilation.

For example, `DoorInSurface_3` in this case uses outdoor air temperatures for ventilation calculation.

Screen Shot 2021-08-31 at 6 42 45 PM

**Comment by xuanluo113** [comment]
> For the updated functions, have you looked at how often consecutive calls use the same argument values? If this happens fairly often, it might make sense to store the previous call argument values and result and check if the new call uses the same argument values.

@JasonGlazer Are you talking about the `SetWindSpeedAt`, `SetSurfacetempAt` functions? Yes, identifying consecutive calls with the same arguments could be an easier cache, but since zones and surfaces are note sorted in height order, the miss rate could be high. I thought Neal's team is working on reducing convective coefficient calculations by grouping surfaces (@amirroth is it true?), and the techniques would be able to apply here.

**Comment by amirroth** [comment]
> @JasonGlazer Are you talking about the `SetWindSpeedAt`, `SetSurfacetempAt` functions? Yes, identifying consecutive calls with the same arguments could be an easier cache, but since zones and surfaces are note sorted in height order, the miss rate could be high. I thought Neal's team is working on reducing convective coefficient calculations by grouping surfaces (@amirroth is it true?), and the techniques would be able to apply here.

The Neal reference doesn't ring a bell. I still think the way to do this is to create a vector of vectors of surface indices where each list contains surfaces of "roughly" the same height, then calculate wind speed once for each interior list and assign to all surfaces within that list. What percentage of runtime does this part of the code take right now? Trying to determine if that is worth doing ...

**Comment by xuanluo113** [comment]
> > @JasonGlazer Are you talking about the `SetWindSpeedAt`, `SetSurfacetempAt` functions? Yes, identifying consecutive calls with the same arguments could be an easier cache, but since zones and surfaces are note sorted in height order, the miss rate could be high. I thought Neal's team is working on reducing convective coefficient calculations by grouping surfaces (@amirroth is it true?), and the techniques would be able to apply here.
> 
> The Neal reference doesn't ring a bell. I still think the way to do this is to create a vector of vectors of surface indices where each list contains surfaces of "roughly" the same height, then calculate wind speed once for each interior list and assign to all surfaces within that list. What percentage of runtime does this part of the code take right now? Trying to determine if that is worth doing ...

@amirroth I don't remember where I got that idea then. It may not be worth it for the surface outdoor condition calculation (run time 0.5% for `SetWindSpeedAt` in a RefMediumOffice annual run).

**Comment by amirroth** [comment]
> @amirroth I don't remember where I got that idea then. It may not be worth it for the surface outdoor condition calculation (run time 0.5% for `SetWindSpeedAt` in a RefMediumOffice annual run).

Yeah, that's borderline. There's probably more impactful things for us to look us. Put a pin in this for now.

**Comment by JasonGlazer** [comment]
@xuanluo0113 it might be worth just testing for the the same value on consecutive calls. A lot of models get defined floor by floor so it might end up being common that the height is the same for consecutive surfaces. I'm not sure but it might be worth checking.

**Comment by mjwitte** [comment]
@xuanluo0113 Now there are eso diffs in 3 files: ConvectionAdaptiveSmallOffice, SurfaceZonePropTest_LocalEnv, ZoneCoupledKivaConvectionAdaptiveSmallOffice. This branch is behind develop, but these results are from 12 hours ago, and nothing was merged between your commit and 12 hrs ago (I think).

**Comment by xuanluo113** [comment]
> @xuanluo0113 Now there are eso diffs in 3 files: ConvectionAdaptiveSmallOffice, SurfaceZonePropTest_LocalEnv, ZoneCoupledKivaConvectionAdaptiveSmallOffice. This branch is behind develop, but these results are from 12 hours ago, and nothing was merged between your commit and 12 hrs ago (I think).

That's odd. Testing locally the diff does not reproduce. I think the diff is caused by the change to convective coefficient calculations [of the other PR](https://github.com/NREL/EnergyPlus/pull/9035/files), which was, by the last commit, not merged into this PR.

**Comment by mitchute** [comment]
The convection diffs are likely due to https://github.com/NREL/EnergyPlus/pull/9035 if you haven't updated recently.

**Comment by mjwitte** [comment]
Well, that's what I thought, but #9035 was merged at 9:38am CDT and the CI results claim to be from 5:23am. <sigh> I hate to trigger another round of CI, but I need to know.
@xuanluo0113 Please update this branch one more time.

**Comment by xuanluo113** [comment]
> Well, that's what I thought, but #9035 was merged at 9:38am CDT and the CI results claim to be from 5:23am. I hate to trigger another round of CI, but I need to know.
> @xuanluo0113 Please update this branch one more time.

@mitchute Thanks for confirming, Matt. 
@mjwitte I'll push the merge commit when finishing a few updates addressing Jason's latest comment.

**Comment by xuanluo113** [comment]
> @xuanluo0113 it might be worth just testing for the the same value on consecutive calls. A lot of models get defined floor by floor so it might end up being common that the height is the same for consecutive surfaces. I'm not sure but it might be worth checking.

@JasonGlazer I tested the methods of skipping consecutive calls with the same surface centroid z in the loop. See the commented-out code at [the latest commit](https://github.com/NREL/EnergyPlus/pull/8949/commits/883c6345ba27cfc34f8bc8ce6fc85635d3944260). I used a highrise-apartment building to test, and although the condition checks were able to filter out 90% of the `std::pow` calculations, the run time saving was not quite significant (~5%). This may be due to the extra assignment and condition checks in the loop. And if tested using a smaller model with fewer surfaces per floor, the run time may go up. So I was thinking it may not be worth it then. Please take a look if I captured what you suggested in the latest test code.

**Comment by JasonGlazer** [comment]
@xuanluo0113 thanks for trying out saving the last value. I'm surprised it wasn't much faster but it is more assignments.

**Comment by JasonGlazer** [comment]
It is weird that you are having an issue with two of the files in the CI that I am also having issues with: ConvectionAdaptiveSmallOffice,  ZoneCoupledKivaConvectionAdaptiveSmallOffice on https://github.com/NREL/EnergyPlus/pull/9024. I have also merged develop recently and am still having the same issue but only on the Linux and MacOS tests not on Windows. Please let me know if you or @mjwitte see what is causing this.

**Comment by mjwitte** [comment]
This is the eso diff summary for ConvectionAdaptiveSmallOffice.

```
"Max absolute diff: 985080.0, field: ATTIC_ROOF_SOUTH:Surface Outside Face Convection Classification Index [](TimeStep), time:  01/21  00:10:00, relative: 999"

"Max relative diff: 999, field: ATTIC_ROOF_SOUTH:Surface Outside Face Convection Classification Index [](TimeStep), time:  01/21  00:10:00, absolute: 985080.0"
```
The possible values are:
```
enum class OutConvClass
{
    Invalid = -1,
    WindwardVertWall = 101,
    LeewardVertWall = 102,
    RoofStable = 103,
    RoofUnstable = 104,
};
```
An absolute diff of 985080.0 in the first timestep of the winter day says this variable isn't getting initialized.

**Comment by amirroth** [comment]
> > @xuanluo0113 it might be worth just testing for the the same value on consecutive calls. A lot of models get defined floor by floor so it might end up being common that the height is the same for consecutive surfaces. I'm not sure but it might be worth checking.
> 
> @JasonGlazer I tested the methods of skipping consecutive calls with the same surface centroid z in the loop. See the commented-out code at [the latest commit](https://github.com/NREL/EnergyPlus/pull/8949/commits/883c6345ba27cfc34f8bc8ce6fc85635d3944260). I used a highrise-apartment building to test, and although the condition checks were able to filter out 90% of the `std::pow` calculations, the run time saving was not quite significant (~5%). This may be due to the extra assignment and condition checks in the loop. And if tested using a smaller model with fewer surfaces per floor, the run time may go up. So I was thinking it may not be worth it then. Please take a look if I captured what you suggested in the latest test code.

Right. So the key to effective caching (which is essentially what this is) is for cache lookup and update to be significantly faster than the uncached calculation. `std::pow` is not as fast as cache lookup, but it's not slower by a factor of two (in fact by your calculations, according to your numbers cache lookup/update is 15% faster than just doing `std::pow`). The underlying math here is much worse than it is for the longer psychrometric functions.

Creating lists of (exterior) surfaces that have roughly the same centroid will not have this problem because you will not incur the overhead of the matching checks. But it may still not be worth pursuing given the overall contribution of this piece of code to runtime. Your call.

**Comment by amirroth** [comment]
> This is the eso diff summary for ConvectionAdaptiveSmallOffice.
> 
> ```
> "Max absolute diff: 985080.0, field: ATTIC_ROOF_SOUTH:Surface Outside Face Convection Classification Index [](TimeStep), time:  01/21  00:10:00, relative: 999"
> 
> "Max relative diff: 999, field: ATTIC_ROOF_SOUTH:Surface Outside Face Convection Classification Index [](TimeStep), time:  01/21  00:10:00, absolute: 985080.0"
> ```
> 
> An absolute diff of 985080.0 in the first timestep of the winter day says this variable isn't getting initialized.

999 is a valid ~absolute~ *relative* diff? That also seems strange.

**Comment by mjwitte** [comment]
> 999 is a valid ~absolute~ *relative* diff?

That's a divide by zero indicator. MathDiff might even have logic to write that explicitly if the base value is zero.

**Comment by xuanluo113** [comment]
I just deleted the commented-out code for testing the caching. 

And if the CI turns green this time, this is ready for another review.

Further investigations of caching and naming conventions of the reportable Q variable will go to the next PR.

**Comment by mjwitte** [comment]
CI is happy here, merging.

**Comment by xuanluo113** [comment]
> CI is happy here, merging.

🎉

**Pull request merged**